src/tests/posix/getopt_test.c

```diff
@@ -11,6 +11,8 @@
 #include <util/array.h>
 #include <unistd.h>
 
+extern void getopt_init(void);
```

Comment:
May be it's better to also add
`#include <getopt.h>`

And except this remove from getopt.h
```
/** setup optind and opterr */
extern void getopt_init(void); /* TODO remove this */
```

Reply:
@ThePiGrepper I agree with @anton-bondarev, it's a good idea to remove `getopt_init` from POSIX header and create another header for this function to use when it's required. For example, you can add a new file `getopt.h` to src/compat/posix/util/ and then export it with `@IncludeExport(path="util")`. After you can use it with `#include <util/getopt.h>`.

By the way, getopt is also at the wrong place: it should be in unistd.h, and it's POSIX, while `getopt_long` is GNU extension, and should be in <getopt.h>. So `getopt_long` should be moved out of the posix folder.

Reply:
I see. I have some questions regarding this change.
1) the suggestion implies we move the function getopt_init() itself or just the 'extern void getopt_init(void);' line to this new file at /src/compat/posix/util/?
2) I assume, based on the filename you chose being 'getopt.h', that you meant to move just the single line and keep the actual function in src/compat/posix/util/getopt.c. if that's correct we'd have a new file, src/compat/posix/util/getopt.h, with a single line. is that correct? and if that's so, would it really be ok to put a header file there?  wouldnt it make more sense to add it at src/include/util/ instead?
3) the last suggestion of changing getopt_long to unistd.h is clear enough. I can send a separate PR for that later, or it could be on this one, if you prefer that.

Reply:
Good questions.

1. I think we can move only getopt_init() declaration to something like `util/getopt.h`, because getopt_init() shares the same global variables with getopt(), so I don't know how to extract getopt_init()  to a separate module.
2. We are trying to keep headers side by side with corresponding sources, so src/compat/posix/util/getopt.h would be ok I think. It's a bit messy with src/include and other includes yet, but we are trying do not add new src/include/xxx if it is possible.
3. Hm, yes, you can actually do that in another PR. I think since `getopt.h` is not POSIX, we should move it somewhere (including getopt_long implementation).

@anton-bondarev What do think?

Reply:
UPD.

1. My previous answer is a bit messy, I want to clarify that because of getopt_init() and getopt() are using the same variables, we can keep them in one module, but only move `getopt_init` declaration to `util/getopt.h` (non-posix) and `getopt` declaration to `unistd.h` (posix).

Reply:
Ok, I got most of it. I got one last doubt though. if a file needs to call getopt_init() at some point (for instance src/framework/cmd/core.c, for cmd_exec(), or src/tests/posix/getopt_test.c) how is it calling it ? should it include the new <util/getopt.h> directly(I think that's an ugly option because it would break all third-party functions, where that file doesnt exist at all), or should unistd.h include <util/getopt.h> so that everything works without change from the outside. I think I got my answer after hearing myself, but let me know if you agree(or disagree) with my call.

Reply:
Ok, I added the requested changes in the second commit. However, I want to add an '#include <util/getopt.h> in unistd.h and it doesnt build. I think I need to add a '@IncludeExport(path="util")' to some Mybuild file, however I dont know exactly to which module I need to add this. can anyone assist me with this? what exactly does Include Export do? and what kind of parameter is path?

Reply:
>  should it include the new <util/getopt.h> directly(I think that's an ugly option because it would break all third-party functions, where that file doesnt exist at all), or should unistd.h include <util/getopt.h> so that everything works without change from the outside

I think it should include <util/getopt.h> directly.

Reply:
>  I think I need to add a '@IncludeExport(path="util")' to some Mybuild file, however I dont know exactly to which module I need to add this.

It looks like you should add this header to the module which implements getopt() - src/compat/posix/util/Util.my (getopt module)

Reply:
I see. 

> > I think I need to add a '@IncludeExport(path="util")' to some Mybuild file, however I dont know exactly to which module I need to add this.
> 
> It looks like you should add this header to the module which implements getopt() - src/compat/posix/util/Util.my (getopt module)

I tried doing that, and it's not working. I'm having trouble understanding how does IncludeExport work in general, that's the main reason I cant really grasp the solution to my issue. can anyone help me with it?

Reply:
> At the end the header is being included in the unistd.h file itself.
I think it makes sense (there's a potential pitfall of doing it this way, and it might happen when the getopt module isnt added to the build, as well as any module requiring it. In that situation, that include in unistd.h will not be able to find the file and it will break the build, right?).

Wouldn't it be better to include <util/getopt.h> only in the files where it is actually required? (not directly in unistd.h)

Reply:
> I tried doing that, and it's not working. I'm having trouble understanding how does IncludeExport work in general, that's the main reason I cant really grasp the solution to my issue. can anyone help me with it?

```
@IncludeExport(path=<path>, target_name=<name.h>)
source "foo.h"
```

The one can use it:

```
#include <path/name.h>
```
Note: `<target_name>` is optional, if it is not specified the actual header name (`foo.h`) will be used.

Reply:
> > At the end the header is being included in the unistd.h file itself.
> > I think it makes sense (there's a potential pitfall of doing it this way, and it might happen when the getopt module isnt added to the build, as well as any module requiring it. In that situation, that include in unistd.h will not be able to find the file and it will break the build, right?).
> 
> Wouldn't it be better to include <util/getopt.h> only in the files where it is actually required? (not directly in unistd.h)

you think so? It would probably solve the issue of not having getopt added to the build and breaking unistd.h. ok, will do.

Reply:
Ok, it's mostly done now. However, the first patch, even with the default, unmodified, x86/qemu conf gives out the following warning:
```
Error: modules dependency cycle:
        embox.fs.driver.devfs_old
        embox.fs.rootfs
        embox.fs.syslib.file_system_full
        embox.driver.common

```
It still compiles and apparently it still works, though. I dont know what this error means..

Reply:
> Ok, it's mostly done now. However, the first patch, even with the default, unmodified, x86/qemu conf gives out the following warning:
> 
> ```
> Error: modules dependency cycle:
>         embox.fs.driver.devfs_old
>         embox.fs.rootfs
>         embox.fs.syslib.file_system_full
>         embox.driver.common
> ```
> 
> It still compiles and apparently it still works, though. I dont know what this error means..

This error shows cyclic dependencies between the modules, but yes, it's not critical.

Reply:
I like how it looks now!